distmv/studentst.go

```diff
@@ -0,0 +1,347 @@
+}
+
+// NewStudentsT creates a new StudentsT with the given nu, mu, and sigma
+// parameters.
+//
+// NewStudentsT panics if len(mu) == 0, or if len(mu) != sigma.Symmetric(). If
+// the covariance matrix is not positive-definite, nil is returned and ok is false.
+func NewStudentsT(nu float64, mu []float64, sigma mat64.Symmetric, src *rand.Rand) (dist *StudentsT, ok bool) {
+	if len(mu) == 0 {
+		panic(badZeroDimension)
+	}
+	dim := sigma.Symmetric()
+	if dim != len(mu) {
+		panic(badSizeMismatch)
+	}
+
+	s := &StudentsT{
+		nu:  nu,
+		mu:  make([]float64, dim),
+		dim: dim,
+		src: src,
+	}
+	copy(s.mu, mu)
+
+	ok = s.chol.Factorize(sigma)
+	if !ok {
+		return nil, false
+	}
+	s.lower.LFromCholesky(&s.chol)
+	s.logSqrtDet = 0.5 * s.chol.LogDet()
+	return s, true
+}
+
+// ConditionStudentsT returns the Student's T distribution that is the receiver
+// conditioned on the input evidence, and the success of the operation.
+// The returned Student's T has dimension
+// n - len(observed), where n is the dimension of the original receiver.
+// The dimension order is preserved during conditioning, so if the value
+// of dimension 1 is observed, the returned normal represents dimensions {0, 2, ...}
+// of the original Student's T distribution.
+//
+// ok indicates whether there was a failure during the update. If ok is false
+// the operation failed and dist is not usable.
+// Mathematically this is impossible, but can occur with finite precision arithmetic.
+func (s *StudentsT) ConditionStudentsT(observed []int, values []float64, src *rand.Rand) (dist *StudentsT, ok bool) {
+	if len(observed) == 0 {
+		panic("studentst: no observed value")
+	}
+	if len(observed) != len(values) {
+		panic(badInputLength)
+	}
+
+	for _, v := range observed {
+		if v < 0 || v >= s.dim {
+			panic("studentst: observed value out of bounds")
+		}
+	}
+
+	s.setSigma()
+
+	newNu, newPhi, newK := studentsTConditional(observed, values, s.nu, s.mu, s.sigma)
+	if newPhi == nil {
+		return nil, false
+	}
+
+	return NewStudentsT(newNu, newPhi, newK, src)
+
+}
+
+// studentsTConditional updates a Student's T distribution based on the observed samples
+// (see documentation for the public function). The Gaussian conditional update
+// is treated as a special case when  nu == math.Inf(1).
+func studentsTConditional(observed []int, values []float64, nu float64, mu []float64, sigma mat64.Symmetric) (newNu float64, newMean []float64, newSigma *mat64.SymDense) {
+	dim := len(mu)
+	ob := len(observed)
+
+	unobserved := findUnob(observed, dim)
+
+	unob := len(unobserved)
+	if unob == 0 {
+		panic("stat: all dimensions observed")
+	}
+
+	mu1 := make([]float64, unob)
+	for i, v := range unobserved {
+		mu1[i] = mu[v]
+	}
+	mu2 := make([]float64, ob) // really v - mu2
+	for i, v := range observed {
+		mu2[i] = values[i] - mu[v]
+	}
+
+	var sigma11, sigma22 mat64.SymDense
+	sigma11.SubsetSym(sigma, unobserved)
+	sigma22.SubsetSym(sigma, observed)
+
+	sigma21 := mat64.NewDense(ob, unob, nil)
```

Comment:
I needed to do something similar for the cca stuff, and I ended up making a symView helper, might be worth considering having something similar somewhere which could be used here?